sdk/src/main/java/io/radar/sdk/util/RadarLogBuffer.kt

```diff
@@ -0,0 +1,199 @@
+) {
+
+    private companion object {
+        const val DIRECTORY = "RadarLogs"
+
+        /**
+         * Maximum number of logs in the buffer.
+         */
+        const val LOG_BUFFER_SIZE_LIMIT = 500
+    }
+
+    /**
+     * Lock for handling changes to [currentFile]
+     */
+    private val fileLock = ReentrantLock()
+
+    /**
+     * Lock for handling changes to the files within [DIRECTORY]
+     */
+    private val directoryLock = ReentrantLock()
+
+    /**
+     * Contains the name of the current log file
+     */
+    private var currentFile: String
+
+    /**
+     * Number of log entries in the current log file.
+     */
+    val size: AtomicInteger
+
+    init {
+        //First find current file
+        val files = getDirectory().listFiles()
+        if (files?.isNotEmpty() == true) {
+            //This gets the latest file that was in use if it can, otherwise it returns the first one.
+            currentFile = files.maxByOrNull { getLastModified(it) }!!.name
+            //This is initialized to the number of lines in the current file
+            size = AtomicInteger(FileInputStream(File(getDirectory(), currentFile)).bufferedReader().readLines().size)
+            if (files.size > 2) {
+                //This case likely means the app started and ended several times without completing a purge. To prevent
+                //excessive log files, purge them now.
+                purge()
+            }
+        } else {
+            //If no such file exists yet, create one.
+            currentFile = getNewFilePath()
+            size = AtomicInteger()
+        }
+    }
+
+    /**
+     * Write a log to the buffer
+     */
+    fun write(level: Radar.RadarLogLevel, message: String) {
+        executor.submit {
+            if (isFull()) {
+                purge()
+            }
+            //Locks file operations while writing to the currentFile
+            fileOp {
+                FileOutputStream(File(getDirectory(), currentFile), true).use { fileOutputStream ->
+                    val radarLog = RadarLog(level, message).toJson().toString()
+                    fileOutputStream.write("$radarLog\n".toByteArray())
+                }
+                size.getAndIncrement()
+            }
+        }
+    }
+
+    /**
+     * Get the logs from the buffer. These are given as a [Flushable] so that a successful callback can cleanup this
+     * log buffer by deleting old log files.
+     */
+    fun getLogs(): Flushable<RadarLog> {
+        val (files, list) = directoryOp {
+            val files = getDirectory().listFiles()
+            val list = mutableListOf<RadarLog>()
+            //Now iterate over the old list of files
+            files?.forEach { file ->
+                FileInputStream(file).bufferedReader().forEachLine { line ->
+                    list.add(RadarLog.fromJson(JSONObject(line)))
+                }
+            }
+            files to list
```

Comment:
is this valid syntax?

Reply:
Yep! it's returning a Pair, which is _destructured_ into the declared values with the same names.

Reply:
neat!